addons/knowledge/controllers/web_unsplash.py

```diff
@@ -0,0 +1,63 @@
+import mimetypes
+import requests
+
+from odoo.addons.web_unsplash.controllers import main
+from odoo import http, tools
+from odoo.http import request
+from odoo.tools.mimetypes import guess_mimetype
+from werkzeug.urls import url_encode
+
+class Web_Unsplash(main.Web_Unsplash):
+
+    @http.route("/web_unsplash/save_random", type='json', auth="user")
+    def save_unsplash_random(self, **kwargs):
+
+        def slugify(s):
+            ''' Keeps only alphanumeric characters, hyphens and spaces from a string.
+                The string will also be truncated to 1024 characters max.
+                :param s: the string to be filtered
+                :return: the sanitized string
+            '''
+            return "".join([c for c in s if c.isalnum() or c in list("- ")])[:1024]
+
+        # Fetch a random image
+        access_key = self._get_access_key()
+        app_id = self.get_unsplash_app_id()
+        if not access_key or not app_id:
+            return {'error': 'key_not_found'}
+        kwargs['client_id'] = access_key
+        req = requests.get('https://api.unsplash.com/photos/random', params=url_encode(kwargs))
+        if req.status_code != requests.codes.ok:
+            return {'error': req.status_code}
+        image_info = req.json()
+
+        # Fetch binary image
+        req = requests.get(image_info['urls']['regular'])
+        if req.status_code != requests.codes.ok:
+            return {'error': req.status_code}
+
+        # Save image in attachments
+        image = tools.image_process(req.content, verify_resolution=True)
+        mimetype = guess_mimetype(image)
+        query = slugify(kwargs.get('query')) + mimetypes.guess_extension(mimetype) or ''
+
+        Attachments = request.env['ir.attachment']
+        res_model = kwargs.get('res_model', 'ir.ui.view')
+        if res_model != 'ir.ui.view' and kwargs.get('res_id'):
+            res_id = int(kwargs['res_id'])
+        else:
+            res_id = None
+        attachment = Attachments.create({
+            'name': '_'.join(['unsplash', image_info['id'], query]),
+            'url': '/' + '/'.join(['unsplash', image_info['id'], query]),
+            'mimetype': mimetype,
+            'raw': image,
+            'public': 'res_model' == 'ir.ui.view',
```

Comment:
I made the same reflexion and concluded that it would maybe be better to do it in another task, because it does not only concern the cover and I think it is worth defining clear specs before doing it. However I prepared the code for this to be done easily.

For example, I'm not sure what will happen if you associate attachments with articles. If you add images to an article, duplicate it, and restrict the access on the original article. The images of the duplicate article will be associated to the id of the original article, so if we lose access to this original article, will we still be able to access them in the duplicated one ? If yes, that means the restricted user could access images from the original article added after that he has been restricted.

In any case things needs to be changed, for the cover and for the body, as for the moment every image that is uploaded in knowledge (cover or body) is public, and anybody could try random id's in '/web/image/id' to get an image.


Note: Images from the media library are saved with "res_model=ir.ui.view" (hardcoded), so in any case for these I would need to extend the associated method. Or we could simply remove images from the media library in the MediaDialog in knowledge